{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5a057443",
   "metadata": {},
   "source": [
    "# CellGuide data pipeline prototype\n",
    "\n",
    "This assumes a local snapshot has been downloaded to the same location as this notebook.\n",
    "\n",
    "Here is a one-liner for downloading the latest prod snapshot:\n",
    "\n",
    "```\n",
    "AWS_PROFILE=single-cell-prod aws s3 sync s3://cellxgene-wmg-prod/$(AWS_PROFILE=single-cell-prod aws s3 cp s3://cellxgene-wmg-prod/latest_snapshot_identifier -)/ prod-snapshot/\n",
    "```\n",
    "\n",
    "**This file should be in the root folder of the `single-cell-data-portal` repo**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8ed8504",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Owlready2 * Warning: optimized Cython parser module 'owlready2_optimized' is not available, defaulting to slower Python implementation\n"
     ]
    }
   ],
   "source": [
    "from backend.wmg.api.v1 import build_filter_dims_values\n",
    "from backend.wmg.data.ontology_labels import ontology_term_label, ontology_term_id_labels\n",
    "from backend.wmg.data.snapshot import WmgSnapshot\n",
    "from backend.wmg.data.query import WmgFiltersQueryCriteria\n",
    "import tiledb\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "import os\n",
    "import openai\n",
    "from backend.wmg.data.utils import get_datasets_from_curation_api, get_collections_from_curation_api\n",
    "from backend.wmg.data.rollup import rollup_across_cell_type_descendants, rollup_across_cell_type_descendants_array\n",
    "import glob\n",
    "import requests\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dd8d69b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sn = \"prod-snapshot\"\n",
    "snapshot = WmgSnapshot(snapshot_identifier=\"\",\n",
    "    expression_summary_cube=tiledb.open(f'{sn}/expression_summary'),\n",
    "    marker_genes_cube=tiledb.open(f'{sn}/marker_genes'),\n",
    "    expression_summary_default_cube=tiledb.open(f'{sn}/expression_summary_default'),\n",
    "    expression_summary_fmg_cube=tiledb.open(f'{sn}/expression_summary_fmg'),                       \n",
    "    cell_counts_cube=tiledb.open(f'{sn}/cell_counts'),\n",
    "    cell_type_orderings=pd.read_json(f'{sn}/cell_type_orderings.json'),\n",
    "    primary_filter_dimensions=json.load(open(f'{sn}/primary_filter_dimensions.json','r')),\n",
    "    dataset_to_gene_ids=json.load(open(f'{sn}/dataset_to_gene_ids.json','r')), \n",
    "    filter_relationships=json.load(open(f'{sn}/filter_relationships.json','r')))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac603523",
   "metadata": {},
   "source": [
    "## Generate all cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30814f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cell_types = [{k: ontology_term_label(k)} for k in ontology_term_id_labels if k.startswith('CL:')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35837527",
   "metadata": {},
   "outputs": [],
   "source": [
    "json.dump(all_cell_types,open('allCellTypes.json','w'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83e6b389",
   "metadata": {},
   "source": [
    "## Generate cell type descriptions using GPT 3.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28e97b2a",
   "metadata": {},
   "source": [
    "System role:\n",
    "> You are a knowledgeable cell biologist that has professional experience writing and curating accurate and informative descriptions of cell types.\n",
    "\n",
    "User role:\n",
    "> I am making a knowledge-base about cell types. Each cell type is a term from the Cell Ontology and will have its own page with a detailed description of that cell type and its function. Please write me a description for \"{cell_type_name}\". Please return only the description and no other dialogue. The description should include information about the cell type's function. The description should be at least three paragraphs long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2653331",
   "metadata": {},
   "outputs": [],
   "source": [
    "openai.organization = \"org-4kBCayJVUBGqH42cJhzZYQ6o\"\n",
    "openai.api_key = \"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\"\n",
    "openai.Model.list()\n",
    "\n",
    "cell_type_descriptions = {}\n",
    "for cell_type in all_cell_types:\n",
    "    cid = list(cell_type.keys())[0]\n",
    "    cname = list(cell_type.values())[0]\n",
    "    print(cid, cname)\n",
    "    \n",
    "    succeeded=False\n",
    "    while not succeeded:\n",
    "        try:\n",
    "            result = openai.ChatCompletion.create(\n",
    "                model=\"gpt-3.5-turbo\",\n",
    "                messages=[\n",
    "                    {\"role\": \"system\", \"content\": \"You are a knowledgeable cell biologist that has professional experience writing and curating accurate and informative descriptions of cell types.\"},\n",
    "                    {\"role\": \"user\", \"content\": f\"I am making a knowledge-base about cell types. Each cell type is a term from the Cell Ontology and will have its own page with a detailed description of that cell type and its function. Please write me a description for \\\"{cname}\\\". Please return only the description and no other dialogue. The description should include information about the cell type's function. The description should be at least three paragraphs long.\"},\n",
    "                ]\n",
    "            ) \n",
    "            succeeded=True\n",
    "        except:\n",
    "            print(\"Trying again due to RLE\")\n",
    "            \n",
    "    print(result['choices'][0]['message']['content'])\n",
    "    \n",
    "    cell_type_descriptions[cid] = result['choices'][0]['message']['content']\n",
    "    \n",
    "    # dump at every iteration to save place\n",
    "    json.dump(cell_type_descriptions,open('allCellTypeDescriptions.json','w'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5747676",
   "metadata": {},
   "source": [
    "## Generate source data per cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "787609a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: DEPLOYMENT_STAGE=test\n"
     ]
    }
   ],
   "source": [
    "%env DEPLOYMENT_STAGE=test\n",
    "\n",
    "def get_title_from_doi(doi):\n",
    "    url = f\"https://api.crossref.org/works/{doi}\"\n",
    "\n",
    "    # Send a GET request to the API\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # If the GET request is successful, the status code will be 200\n",
    "    if response.status_code == 200:\n",
    "        # Get the response data\n",
    "        data = response.json()\n",
    "\n",
    "        # Get the title and citation count from the data\n",
    "        try:\n",
    "            title = data['message']['title'][0]\n",
    "        except:\n",
    "            try:\n",
    "                title = data['message']['items'][0]['title'][0]\n",
    "            except:\n",
    "                return doi\n",
    "        return title\n",
    "    else:\n",
    "        return doi\n",
    "    \n",
    "def format_citation(metadata):\n",
    "    first_author = metadata['publisher_metadata']['authors'][0]\n",
    "    if \"family\" in first_author:\n",
    "        author_str = f\"{first_author['family']}, {first_author['given']} et al.\"\n",
    "    else:\n",
    "        author_str = f\"{first_author['name']} et al.\"\n",
    "    \n",
    "    journal = metadata['publisher_metadata']['journal']\n",
    "    year = metadata['publisher_metadata']['published_year']\n",
    "    \n",
    "    return f\"{author_str} ({year}) {journal}\"\n",
    "\n",
    "snapshot.build_dataset_metadata_dict()\n",
    "\n",
    "datasets = get_datasets_from_curation_api()\n",
    "collections = get_collections_from_curation_api()\n",
    "\n",
    "collections_dict = {collection['collection_id']: collection for collection in collections}\n",
    "datasets_dict = {dataset['dataset_id']: dataset for dataset in datasets}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "40d0a686",
   "metadata": {},
   "outputs": [],
   "source": [
    "cts = [list(i.keys())[0] for i in all_cell_types]\n",
    "\n",
    "DATA = {}\n",
    "for i in cts:\n",
    "    criteria = WmgFiltersQueryCriteria(organism_ontology_term_id=\"NCBITaxon:9606\",\n",
    "                                       **dict(cell_type_ontology_term_ids=[i]))\n",
    "    res = build_filter_dims_values(criteria, snapshot)\n",
    "    datasets = res['datasets']\n",
    "    collections_to_datasets = {}\n",
    "    for dataset in datasets:\n",
    "        dataset = datasets_dict[dataset['id']]\n",
    "        \n",
    "        a = collections_to_datasets.get(dataset['collection_id'],{})\n",
    "        \n",
    "        a['collection_name'] = collections_dict[dataset['collection_id']]['name']\n",
    "        a['collection_url'] = collections_dict[dataset['collection_id']]['collection_url']\n",
    "        a['publication_url'] = collections_dict[dataset['collection_id']]['doi']\n",
    "        if collections_dict[dataset['collection_id']]['publisher_metadata']:\n",
    "            a['publication_title'] = format_citation(collections_dict[dataset['collection_id']])\n",
    "        else:\n",
    "            a['publication_title'] = \"Publication\"\n",
    "        a['tissue'] = dataset['tissue']\n",
    "        a['disease'] = dataset['disease']\n",
    "        a['organism'] = dataset['organism']\n",
    "        \n",
    "        collections_to_datasets[dataset['collection_id']]=a\n",
    "    \n",
    "    DATA[i] = list(collections_to_datasets.values())                                 \n",
    "    \n",
    "    # dump at every iteration to save place\n",
    "    json.dump(DATA,open('allSourceData.json','w'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "069bef8e",
   "metadata": {},
   "source": [
    "## Generate enriched genes and expression metrics per cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0161691b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_counts_df = snapshot.cell_counts_cube.df[:]\n",
    "cell_counts_df = cell_counts_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id']).sum(numeric_only=True)\n",
    "\n",
    "expressions_df = snapshot.expression_summary_default_cube.df[:]\n",
    "expressions_df = expressions_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']).sum(numeric_only=True)\n",
    "expressions_df = expressions_df.reset_index()\n",
    "\n",
    "all_cell_type_ids = [list(i.keys())[0] for i in all_cell_types]\n",
    "\n",
    "index = pd.Index(list(itertools.product(cell_counts_df.index.get_level_values('organism_ontology_term_id').unique(),all_cell_type_ids)))\n",
    "index = index.set_names(['organism_ontology_term_id','cell_type_ontology_term_id'])\n",
    "universe_cell_counts_df = pd.DataFrame(index = index)\n",
    "universe_cell_counts_df['n_cells']=0\n",
    "universe_cell_counts_df['n_cells'][cell_counts_df.index] = cell_counts_df['n_cells']\n",
    "\n",
    "universe_cell_counts_df = rollup_across_cell_type_descendants(\n",
    "    universe_cell_counts_df.reset_index()\n",
    ")\n",
    "universe_cell_counts_df=universe_cell_counts_df[universe_cell_counts_df['n_cells']>0]\n",
    "universe_cell_counts_df = universe_cell_counts_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id']).sum()\n",
    "\n",
    "x = list(zip(*expressions_df[['organism_ontology_term_id','cell_type_ontology_term_id']].values.T))\n",
    "y = list(expressions_df['gene_ontology_term_id'])\n",
    "\n",
    "xu = list(set(x))\n",
    "yu = list(set(y))\n",
    "\n",
    "x_index = pd.Series(index=pd.Index(xu),data=np.arange(len(xu)))\n",
    "y_index = pd.Series(index=pd.Index(yu),data=np.arange(len(yu)))\n",
    "\n",
    "e_nnz = np.zeros((len(xu),len(yu)))\n",
    "e_sum = np.zeros((len(xu),len(yu)))\n",
    "\n",
    "e_nnz[x_index[x].values, y_index[y].values] = expressions_df['nnz']\n",
    "e_sum[x_index[x].values, y_index[y].values] = expressions_df['sum']\n",
    "\n",
    "available_combinations = set(list(universe_cell_counts_df.index.values))\n",
    "missing_combinations = available_combinations.difference(xu)\n",
    "\n",
    "xu = xu + list(missing_combinations)\n",
    "x_index = pd.Series(index=pd.Index(xu),data=np.arange(len(xu)))\n",
    "e_nnz = np.vstack((e_nnz,np.zeros((len(missing_combinations),e_nnz.shape[1]))))\n",
    "e_sum = np.vstack((e_sum,np.zeros((len(missing_combinations),e_sum.shape[1]))))\n",
    "organisms = x_index.index.get_level_values(0)\n",
    "organisms_u = list(set(organisms))\n",
    "cell_types = x_index.index.get_level_values(1)\n",
    "e_nnz_rollup = np.zeros_like(e_nnz)\n",
    "e_sum_rollup = np.zeros_like(e_sum)\n",
    "for organism in organisms_u:\n",
    "    cell_types_o = cell_types[organisms==organism]\n",
    "    e_nnz_o = e_nnz[organisms==organism]\n",
    "    e_sum_o = e_sum[organisms==organism]    \n",
    "    e_nnz_o = rollup_across_cell_type_descendants_array(e_nnz_o,cell_types_o)\n",
    "    e_sum_o = rollup_across_cell_type_descendants_array(e_sum_o,cell_types_o)  \n",
    "    e_nnz_rollup[organisms==organism]=e_nnz_o\n",
    "    e_sum_rollup[organisms==organism]=e_sum_o"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75f24d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cells = universe_cell_counts_df['n_cells'][x_index.index]\n",
    "n_cells = np.tile(n_cells.values[:,None],(1,e_nnz_rollup.shape[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d05c991",
   "metadata": {},
   "outputs": [],
   "source": [
    "r = snapshot.expression_summary_fmg_cube.df[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "a32facbe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "7527fb15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "28bb88e3",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[177], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28;43mset\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m)\u001b[49m))\u001b[38;5;241m==\u001b[39m\u001b[38;5;28mlen\u001b[39m(index)\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/base.py:780\u001b[0m, in \u001b[0;36mIndexOpsMixin.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    768\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    769\u001b[0m \u001b[38;5;124;03mReturn an iterator of the values.\u001b[39;00m\n\u001b[1;32m    770\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    777\u001b[0m \u001b[38;5;124;03miterator\u001b[39;00m\n\u001b[1;32m    778\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    779\u001b[0m \u001b[38;5;66;03m# We are explicitly making element iterators.\u001b[39;00m\n\u001b[0;32m--> 780\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_values\u001b[49m, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m    781\u001b[0m     \u001b[38;5;66;03m# Check type instead of dtype to catch DTA/TDA\u001b[39;00m\n\u001b[1;32m    782\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28miter\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values)\n\u001b[1;32m    783\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/_libs/properties.pyx:36\u001b[0m, in \u001b[0;36mpandas._libs.properties.CachedProperty.__get__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/indexes/multi.py:748\u001b[0m, in \u001b[0;36mMultiIndex._values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    745\u001b[0m     vals \u001b[38;5;241m=\u001b[39m algos\u001b[38;5;241m.\u001b[39mtake_nd(vals, codes, fill_value\u001b[38;5;241m=\u001b[39mindex\u001b[38;5;241m.\u001b[39m_na_value)\n\u001b[1;32m    746\u001b[0m     values\u001b[38;5;241m.\u001b[39mappend(vals)\n\u001b[0;32m--> 748\u001b[0m arr \u001b[38;5;241m=\u001b[39m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfast_zip\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    749\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m arr\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "bddb128d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "a80c620c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "2b7c98a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "c7e9c6c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "a68592a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "d05a61a4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "cc2f4456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCBITaxon:9825\n",
      "NCBITaxon:9544\n",
      "NCBITaxon:10090\n",
      "NCBITaxon:9606\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "c6c24022",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "2d5f1c58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "e7b94792",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "7a4de340",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472059f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "839775ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "bda91810",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[105], line 12\u001b[0m\n\u001b[1;32m     10\u001b[0m entries_to_add \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[0;32m---> 12\u001b[0m     expressions_df_per_gene \u001b[38;5;241m=\u001b[39m expressions_df[\u001b[43mexpressions_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_ontology_term_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m]\n\u001b[1;32m     13\u001b[0m     available_combinations_per_gene \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(\u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39mexpressions_df_per_gene[group_by_terms]\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mT))\n\u001b[1;32m     15\u001b[0m     missing_combinations \u001b[38;5;241m=\u001b[39m available_combinations\u001b[38;5;241m.\u001b[39mdifference(available_combinations_per_gene)\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/ops/common.py:81\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     77\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[1;32m     79\u001b[0m other \u001b[38;5;241m=\u001b[39m item_from_zerodim(other)\n\u001b[0;32m---> 81\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mother\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/arraylike.py:40\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__eq__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m---> 40\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cmp_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meq\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/series.py:6097\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   6094\u001b[0m rvalues \u001b[38;5;241m=\u001b[39m extract_array(other, extract_numpy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, extract_range\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m   6096\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m np\u001b[38;5;241m.\u001b[39merrstate(\u001b[38;5;28mall\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m-> 6097\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcomparison_op\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6099\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_result(res_values, name\u001b[38;5;241m=\u001b[39mres_name)\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/ops/array_ops.py:286\u001b[0m, in \u001b[0;36mcomparison_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    283\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[1;32m    285\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_object_dtype(lvalues\u001b[38;5;241m.\u001b[39mdtype) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(rvalues, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 286\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mcomp_method_OBJECT_ARRAY\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    288\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    289\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniconda/envs/scdp2/lib/python3.9/site-packages/pandas/core/ops/array_ops.py:75\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[0;34m(op, x, y)\u001b[0m\n\u001b[1;32m     73\u001b[0m     result \u001b[38;5;241m=\u001b[39m libops\u001b[38;5;241m.\u001b[39mvec_compare(x\u001b[38;5;241m.\u001b[39mravel(), y\u001b[38;5;241m.\u001b[39mravel(), op)\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 75\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mlibops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscalar_compare\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mravel\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\u001b[38;5;241m.\u001b[39mreshape(x\u001b[38;5;241m.\u001b[39mshape)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "group_by_terms = list(universe_cell_counts_df.index.names)\n",
    "genes = list(set(expressions_df[\"gene_ontology_term_id\"]))\n",
    "\n",
    "numeric_columns = list(\n",
    "    expressions_df.columns[[np.issubdtype(dtype, np.number) for dtype in expressions_df.dtypes]]\n",
    ")\n",
    "\n",
    "\n",
    "entries_to_add = []\n",
    "for gene in genes:\n",
    "    expressions_df_per_gene = expressions_df[expressions_df[\"gene_ontology_term_id\"] == gene]\n",
    "    available_combinations_per_gene = set(zip(*expressions_df_per_gene[group_by_terms].values.T))\n",
    "\n",
    "    missing_combinations = available_combinations.difference(available_combinations_per_gene)\n",
    "    for combo in missing_combinations:\n",
    "        entry = {dim: combo[i] for i, dim in enumerate(group_by_terms)}\n",
    "        entry.update({col: 0 for col in numeric_columns})\n",
    "        entry[\"gene_ontology_term_id\"] = gene\n",
    "        entries_to_add.append(entry)\n",
    "expressions_df = pd.concat((expressions_df, pd.DataFrame(entries_to_add)), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58a87d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "expressions_df_rollup = rollup_across_cell_type_descendants(\n",
    "    expressions_df.groupby(['organism_ontology_term_id','gene_ontology_term_id','cell_type_ontology_term_id']).sum(numeric_only=True).reset_index()\n",
    ")\n",
    "\n",
    "cell_counts_df_rollup = universe_cell_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70ad8197",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b92c6c1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2510cbeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_df = snapshot.marker_genes_cube.df[:]\n",
    "markers_df=markers_df[markers_df['effect_size_ttest']>0]\n",
    "markers_df=markers_df[markers_df['p_value_ttest']<1e-5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0a491ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_df_agg = markers_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']).agg({'effect_size_ttest': 'max'}).reset_index()\n",
    "markers_df_agg2 = markers_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']).agg({'effect_size_ttest': 'mean'}).reset_index()\n",
    "markers_df_agg3 = markers_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']).agg({'effect_size_ttest': 'min'}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d4c0986",
   "metadata": {},
   "outputs": [],
   "source": [
    "# top_per_group = markers_df_agg.groupby(['organism_ontology_term_id','cell_type_ontology_term_id']).apply(lambda x: x.nlargest(200, 'effect_size_ttest'))\n",
    "\n",
    "# columns = ['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']\n",
    "# o_ct_genes = list(zip(*top_per_group[columns].values.T))\n",
    "\n",
    "expressions_df = snapshot.expression_summary_default_cube.df[:]\n",
    "\n",
    "cell_counts_df = snapshot.cell_counts_cube.df[:]\n",
    "\n",
    "cell_counts_df_rollup = rollup_across_cell_type_descendants(\n",
    "    cell_counts_df.groupby(['organism_ontology_term_id','cell_type_ontology_term_id']).sum(numeric_only=True).reset_index()\n",
    ")\n",
    "\n",
    "expressions_df_rollup = rollup_across_cell_type_descendants(\n",
    "    expressions_df.groupby(['organism_ontology_term_id','gene_ontology_term_id','cell_type_ontology_term_id']).sum(numeric_only=True).reset_index()\n",
    ")\n",
    "\n",
    "filt1 = expressions_df_rollup['cell_type_ontology_term_id'].isin(top_per_group['cell_type_ontology_term_id'].unique())\n",
    "\n",
    "filt2 = expressions_df_rollup['gene_ontology_term_id'].isin(top_per_group['gene_ontology_term_id'].unique())\n",
    "\n",
    "filt = np.logical_and(filt1,filt2)\n",
    "\n",
    "expressions_df_rollup = expressions_df_rollup[filt]\n",
    "\n",
    "expressions_df_rollup.index = pd.Index(list(zip(*expressions_df_rollup[['organism_ontology_term_id','cell_type_ontology_term_id','gene_ontology_term_id']].values.T)))\n",
    "\n",
    "cell_counts_df_rollup = cell_counts_df_rollup.groupby(['organism_ontology_term_id','cell_type_ontology_term_id']).sum()['n_cells']\n",
    "\n",
    "gene_id_to_symbol={}\n",
    "all_genes = []\n",
    "for k in snapshot.primary_filter_dimensions['gene_terms']:\n",
    "    for i in snapshot.primary_filter_dimensions['gene_terms'][k]:\n",
    "        gene_id_to_symbol.update(i)\n",
    "        all_genes.append(list(i.keys())[0])\n",
    "\n",
    "gene_id_to_name = pd.read_csv('ensembl_gene_ids_to_descriptions.tsv',sep='\\t')\n",
    "\n",
    "gene_id_to_name = gene_id_to_name.set_index('Ensembl GeneIDs')['Description'].to_dict()\n",
    "\n",
    "data={}\n",
    "for i in o_ct_genes:\n",
    "    o,ct,gene = i\n",
    "    \n",
    "    nnz = expressions_df_rollup['nnz'][i]\n",
    "    s = expressions_df_rollup['sum'][i]\n",
    "    n_cells = cell_counts_df_rollup[(o,ct)]\n",
    "    \n",
    "    a = data.get(ct,[])\n",
    "    a.append({\n",
    "        'me': s/nnz if nnz > 0 else 0,\n",
    "        'pc': nnz/n_cells,\n",
    "        'symbol': gene_id_to_symbol[gene],\n",
    "        'name': gene_id_to_name.get(gene,gene_id_to_symbol[gene]),\n",
    "        'organism': o\n",
    "    })\n",
    "    data[ct]=a\n",
    "json.dump(data,open('allEnrichedGenes.json','w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "cd420b09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'me': 3.6141178480433385,\n",
       "  'pc': 0.9350511171762964,\n",
       "  'symbol': 'Hexb',\n",
       "  'name': 'hexosaminidase B',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.7239227894658913,\n",
       "  'pc': 0.8591849007725402,\n",
       "  'symbol': 'Lgmn',\n",
       "  'name': 'legumain',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.881655434917185,\n",
       "  'pc': 0.8406809455521118,\n",
       "  'symbol': 'Cx3cr1',\n",
       "  'name': 'chemokine (C-X3-C motif) receptor 1',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.5326666873860435,\n",
       "  'pc': 0.8372577138363325,\n",
       "  'symbol': 'Csf1r',\n",
       "  'name': 'colony stimulating factor 1 receptor',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 3.9119411606282246,\n",
       "  'pc': 0.8158856455567377,\n",
       "  'symbol': 'C1qa',\n",
       "  'name': 'complement component 1, q subcomponent, alpha polypeptide',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 3.739387100763143,\n",
       "  'pc': 0.8183374196234445,\n",
       "  'symbol': 'C1qb',\n",
       "  'name': 'complement component 1, q subcomponent, beta polypeptide',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 3.5894259435138465,\n",
       "  'pc': 0.7884535319424527,\n",
       "  'symbol': 'C1qc',\n",
       "  'name': 'complement component 1, q subcomponent, C chain',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.5006629829439833,\n",
       "  'pc': 0.7465420733681825,\n",
       "  'symbol': 'P2ry12',\n",
       "  'name': 'purinergic receptor P2Y, G-protein coupled 12',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.6637998472914957,\n",
       "  'pc': 0.7327103668409123,\n",
       "  'symbol': 'Trem2',\n",
       "  'name': 'triggering receptor expressed on myeloid cells 2',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.0302845722657494,\n",
       "  'pc': 0.7262339825137624,\n",
       "  'symbol': 'Lpcat2',\n",
       "  'name': 'lysophosphatidylcholine acyltransferase 2',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.399883165865109,\n",
       "  'pc': 0.8079752047000046,\n",
       "  'symbol': 'Ly86',\n",
       "  'name': 'lymphocyte antigen 86',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.2184853658717447,\n",
       "  'pc': 0.7798029328769024,\n",
       "  'symbol': 'Tgfbr1',\n",
       "  'name': 'transforming growth factor, beta receptor I',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 1.9900520526479444,\n",
       "  'pc': 0.7066660498681593,\n",
       "  'symbol': 'Itgb5',\n",
       "  'name': 'integrin beta 5',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.1797443049162832,\n",
       "  'pc': 0.6934819817736041,\n",
       "  'symbol': 'Siglech',\n",
       "  'name': 'sialic acid binding Ig-like lectin H',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 3.509725096481906,\n",
       "  'pc': 0.8270805384650969,\n",
       "  'symbol': 'Ctsd',\n",
       "  'name': 'cathepsin D',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.225780136411834,\n",
       "  'pc': 0.6653097099505019,\n",
       "  'symbol': 'Gpr34',\n",
       "  'name': 'G protein-coupled receptor 34',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 1.925902878750266,\n",
       "  'pc': 0.6522181616320488,\n",
       "  'symbol': 'Hpgds',\n",
       "  'name': 'hematopoietic prostaglandin D synthase',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 1.900570671469404,\n",
       "  'pc': 0.6433362631262433,\n",
       "  'symbol': 'Lair1',\n",
       "  'name': 'leukocyte-associated Ig-like receptor 1',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.1340894879388657,\n",
       "  'pc': 0.7536660961280474,\n",
       "  'symbol': 'Epb41l2',\n",
       "  'name': 'erythrocyte membrane protein band 4.1 like 2',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 1.9666292974141235,\n",
       "  'pc': 0.6033214599620669,\n",
       "  'symbol': 'Aif1',\n",
       "  'name': 'allograft inflammatory factor 1',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.0186027947746696,\n",
       "  'pc': 0.5953647592172827,\n",
       "  'symbol': 'Mafb',\n",
       "  'name': 'MAF bZIP transcription factor B',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.0482750682378494,\n",
       "  'pc': 0.5910625896285331,\n",
       "  'symbol': 'Olfml3',\n",
       "  'name': 'olfactomedin-like 3',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.076457049589202,\n",
       "  'pc': 0.5912013692926863,\n",
       "  'symbol': 'Fcrls',\n",
       "  'name': 'Fc receptor like 2',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.2734895167798777,\n",
       "  'pc': 0.6878382754313734,\n",
       "  'symbol': 'Ctsl',\n",
       "  'name': 'cathepsin L',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 1.991689053167662,\n",
       "  'pc': 0.5421658879585511,\n",
       "  'symbol': 'Tmem119',\n",
       "  'name': 'transmembrane protein 119',\n",
       "  'organism': 'NCBITaxon:10090'},\n",
       " {'me': 2.632558094744509,\n",
       "  'pc': 0.3132019317160827,\n",
       "  'symbol': 'HLA-DRA',\n",
       "  'name': 'major histocompatibility complex, class II, DR alpha',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.0854089464087098,\n",
       "  'pc': 0.16408355795148247,\n",
       "  'symbol': 'HLA-DPB1',\n",
       "  'name': 'major histocompatibility complex, class II, DP beta 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.994312723147366,\n",
       "  'pc': 0.5150915318957772,\n",
       "  'symbol': 'CD74',\n",
       "  'name': 'CD74 molecule',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.346439061390485,\n",
       "  'pc': 0.21747248427672955,\n",
       "  'symbol': 'HLA-DPA1',\n",
       "  'name': 'major histocompatibility complex, class II, DP alpha 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.6066992168339818,\n",
       "  'pc': 0.42931547619047616,\n",
       "  'symbol': 'C1QB',\n",
       "  'name': 'complement C1q B chain',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.4779092046862097,\n",
       "  'pc': 0.30601415094339623,\n",
       "  'symbol': 'C1QA',\n",
       "  'name': 'complement C1q A chain',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.583056129881646,\n",
       "  'pc': 0.38787342767295596,\n",
       "  'symbol': 'C1QC',\n",
       "  'name': 'complement C1q C chain',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.190333731731965,\n",
       "  'pc': 0.27088948787061995,\n",
       "  'symbol': 'TYROBP',\n",
       "  'name': 'transmembrane immune signaling adaptor TYROBP',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.370883847157207,\n",
       "  'pc': 0.2879043126684636,\n",
       "  'symbol': 'HLA-DRB1',\n",
       "  'name': 'major histocompatibility complex, class II, DR beta 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 3.2462048182671674,\n",
       "  'pc': 0.628032345013477,\n",
       "  'symbol': 'SPP1',\n",
       "  'name': 'secreted phosphoprotein 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.171956557679105,\n",
       "  'pc': 0.17381233153638814,\n",
       "  'symbol': 'AIF1',\n",
       "  'name': 'allograft inflammatory factor 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 1.8959913343298735,\n",
       "  'pc': 0.07777403414195867,\n",
       "  'symbol': 'HLA-DQB1',\n",
       "  'name': 'major histocompatibility complex, class II, DQ beta 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.3274489049066114,\n",
       "  'pc': 0.23676156783468105,\n",
       "  'symbol': 'GPX1',\n",
       "  'name': 'glutathione peroxidase 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.200083563502151,\n",
       "  'pc': 0.17948393980233604,\n",
       "  'symbol': 'APOC1',\n",
       "  'name': 'apolipoprotein C1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.320362111828332,\n",
       "  'pc': 0.6035208894878706,\n",
       "  'symbol': 'RNASET2',\n",
       "  'name': 'ribonuclease T2',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.1436586417055166,\n",
       "  'pc': 0.1987449460916442,\n",
       "  'symbol': 'FCER1G',\n",
       "  'name': 'Fc epsilon receptor Ig',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 1.8399249369061759,\n",
       "  'pc': 0.04728212039532794,\n",
       "  'symbol': 'HLA-DQA1',\n",
       "  'name': 'major histocompatibility complex, class II, DQ alpha 1',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 3.189274478876185,\n",
       "  'pc': 0.02517127133872417,\n",
       "  'symbol': 'CCL3',\n",
       "  'name': 'C-C motif chemokine ligand 3',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 1.8312845994156544,\n",
       "  'pc': 0.0912932389937107,\n",
       "  'symbol': 'HLA-DMA',\n",
       "  'name': 'major histocompatibility complex, class II, DM alpha',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 3.1724549515445184,\n",
       "  'pc': 0.5794446316262354,\n",
       "  'symbol': 'FTL',\n",
       "  'name': 'ferritin light chain',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.267850315698625,\n",
       "  'pc': 0.2317918912848158,\n",
       "  'symbol': 'CD14',\n",
       "  'name': 'CD14 molecule',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 1.9657470125546634,\n",
       "  'pc': 0.19582491015274034,\n",
       "  'symbol': 'FCGR3A',\n",
       "  'name': 'Fc gamma receptor IIIa',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.5544859743283337,\n",
       "  'pc': 0.5110343665768194,\n",
       "  'symbol': 'APBB1IP',\n",
       "  'name': 'amyloid beta precursor protein binding family B member 1 interacting protein',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 2.158601108238001,\n",
       "  'pc': 0.16321316262354,\n",
       "  'symbol': 'RPS17',\n",
       "  'name': 'ribosomal protein S17',\n",
       "  'organism': 'NCBITaxon:9606'},\n",
       " {'me': 3.142920065579015,\n",
       "  'pc': 0.5485596361185984,\n",
       "  'symbol': 'ARHGAP24',\n",
       "  'name': 'Rho GTPase activating protein 24',\n",
       "  'organism': 'NCBITaxon:9606'}]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['CL:0000129']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c786ce0",
   "metadata": {},
   "source": [
    "## Generate canonical marker genes data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11182e9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_prefix_cols(prefix, cols):\n",
    "    i=1\n",
    "    prefix_cols = []\n",
    "    while True:\n",
    "        col = f\"{prefix}{i}\"\n",
    "        if col in cols:\n",
    "            prefix_cols.append(col)\n",
    "            i+=1\n",
    "        elif col.upper() in cols:\n",
    "            prefix_cols.append(col.upper())\n",
    "            i+=1            \n",
    "        elif col.lower() in cols:\n",
    "            prefix_cols.append(col.lower())\n",
    "            i+=1\n",
    "        else:\n",
    "            break   \n",
    "    return prefix_cols\n",
    "\n",
    "def get_all_suffix_cols(prefix,suffix, cols):\n",
    "    i=1\n",
    "    suffix_cols = []\n",
    "    while True:\n",
    "        col = f\"{prefix}{i}{suffix}\"\n",
    "        if col in cols:\n",
    "            suffix_cols.append(col)\n",
    "            i+=1\n",
    "        elif col.upper() in cols:\n",
    "            suffix_cols.append(col.upper())\n",
    "            i+=1            \n",
    "        elif col.lower() in cols:\n",
    "            suffix_cols.append(col.lower())\n",
    "            i+=1\n",
    "        else:\n",
    "            break    \n",
    "    return suffix_cols\n",
    "\n",
    "def get_gene_name(gene):\n",
    "    a = requests.get(f\"https://api.cellxgene.dev.single-cell.czi.technology/gene_info/v1/gene_info?gene={gene}\")\n",
    "    if a.status_code == 200:\n",
    "        r = a.json()\n",
    "        return r['name']\n",
    "    else:\n",
    "        return gene\n",
    "    \n",
    "def try_delete(d, k):\n",
    "    try:\n",
    "        del d[k]\n",
    "    except:\n",
    "        try:\n",
    "            del d[k[0]+k[1:].lower()]\n",
    "        except:\n",
    "            pass\n",
    "    \n",
    "def get_title_from_doi(doi):\n",
    "    url = f\"https://api.crossref.org/works/{doi}\"\n",
    "\n",
    "    # Send a GET request to the API\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # If the GET request is successful, the status code will be 200\n",
    "    if response.status_code == 200:\n",
    "        # Get the response data\n",
    "        data = response.json()\n",
    "\n",
    "        # Get the title and citation count from the data\n",
    "        try:\n",
    "            title = data['message']['title'][0]\n",
    "        except:\n",
    "            try:\n",
    "                title = data['message']['items'][0]['title'][0]\n",
    "            except:\n",
    "                return doi\n",
    "        return title\n",
    "    else:\n",
    "        return doi\n",
    "    \n",
    "\n",
    "\n",
    "def get_tissue_name(t):\n",
    "    t=t.replace(':','_')\n",
    "    urls = [\n",
    "        f\"https://www.ebi.ac.uk/ols4/api/ontologies/clo/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252F{t}\",\n",
    "        f\"https://www.ebi.ac.uk/ols4/api/ontologies/envo/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252F{t}\",\n",
    "        f\"https://www.ebi.ac.uk/ols4/api/ontologies/flopo/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252F{t}\",\n",
    "        f\"https://www.ebi.ac.uk/ols4/api/ontologies/doid/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252F{t}\",\n",
    "    ]\n",
    "    for url in urls:    \n",
    "        response = requests.get(url)\n",
    "        if response.status_code==200:\n",
    "            r = response.json()\n",
    "            return r['label']\n",
    "    return t\n",
    "\n",
    "pf = json.load(open('dev-sn/primary_filter_dimensions.json','r'))\n",
    "all_human_genes = [list(i.values())[0] for i in pf['gene_terms']['NCBITaxon:9606']]\n",
    "\n",
    "X = tiledb.open('dev-sn/cell_counts/')\n",
    "cc = X.df[:]\n",
    "tissue_original = list(set(cc['tissue_original_ontology_term_id']))\n",
    "tissue = list(set(cc['tissue_ontology_term_id']))\n",
    "\n",
    "m = {}\n",
    "[m.update(i) for i in pf['tissue_terms']['NCBITaxon:9606']];\n",
    "for i in tissue:\n",
    "    if i not in m:\n",
    "        m[i]=i\n",
    "        \n",
    "for i in tissue_original:\n",
    "    if i not in m:\n",
    "        m[i]=i\n",
    "\n",
    "files = glob.glob('tables/*.csv')\n",
    "\n",
    "parsed_table_entries = []\n",
    "\n",
    "seen=[]\n",
    "for file in files:\n",
    "    print(file)\n",
    "\n",
    "    df = pd.read_csv(file,skiprows=10)\n",
    "    assert df.columns[0]=='AS/1'\n",
    "\n",
    "    cols = list(df.columns)\n",
    "\n",
    "    ref_prefix = \"Ref/\"\n",
    "    ref_doi_suffix = \"/DOI\"\n",
    "\n",
    "    ref_prefixes = get_all_prefix_cols(ref_prefix,cols)\n",
    "    if len(ref_prefixes):\n",
    "        prefix=ref_prefixes[0].split('/')[0]+\"/\"\n",
    "        ref_suffixes=get_all_suffix_cols(prefix,ref_doi_suffix,cols)\n",
    "        ref_suffixes_notes=get_all_suffix_cols(prefix,\"/NOTES\",cols)\n",
    "    else:\n",
    "        ref_suffixes=[]\n",
    "        ref_suffixes_notes=[]\n",
    "\n",
    "\n",
    "    gene_prefix = \"BGene/\"\n",
    "    gene_label_suffix = \"/LABEL\"\n",
    "\n",
    "    gene_prefixes = get_all_prefix_cols(gene_prefix,cols)\n",
    "    if len(gene_prefixes):\n",
    "        prefix=gene_prefixes[0].split('/')[0]+\"/\"\n",
    "        gene_suffixes=get_all_suffix_cols(prefix,gene_label_suffix,cols)\n",
    "    else:\n",
    "        gene_suffixes=[]\n",
    "\n",
    "    tissue_prefix = \"AS/\"\n",
    "    protein_label_suffix = \"/ID\"\n",
    "\n",
    "    tissue_prefixes = get_all_prefix_cols(tissue_prefix,cols)\n",
    "    if len(tissue_prefixes):\n",
    "        prefix=tissue_prefixes[0].split('/')[0]+\"/\"\n",
    "        tissue_suffixes=get_all_suffix_cols(prefix,protein_label_suffix,cols) \n",
    "    else:\n",
    "        tissue_suffixes=[]\n",
    "\n",
    "    ct = \"CT/1\"\n",
    "    ctid = \"CT/1/ID\"\n",
    "    try:\n",
    "        assert len(ref_prefixes) > 0\n",
    "        assert len(ref_suffixes) > 0\n",
    "        assert len(gene_prefixes) > 0 \n",
    "        assert len(gene_suffixes) > 0 \n",
    "        assert len(tissue_suffixes) > 0\n",
    "    except:\n",
    "        print(\"Skipping\")\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    for n in range(df.shape[0]):\n",
    "        res = df.iloc[n].to_dict()\n",
    "        data_tmp = {i: res[i] for i in res if i in [ct, ctid] + ref_prefixes+ref_suffixes+gene_prefixes+gene_suffixes+tissue_suffixes+ref_suffixes_notes}\n",
    "        data = data_tmp.copy()\n",
    "        genes = []\n",
    "        gene_to_key = {}\n",
    "        for i in gene_prefixes:\n",
    "            data_tmp[i] = str(data_tmp[i]).split(' ')[0]\n",
    "            if data_tmp[i].upper() not in all_human_genes or data_tmp[i]=='nan':\n",
    "                try_delete(data,i)\n",
    "                try_delete(data,i+'/LABEL')\n",
    "            else:\n",
    "                data[i] = data_tmp[i].upper()\n",
    "                genes.append(data[i])\n",
    "                gene_to_key[data[i]] = i\n",
    "\n",
    "        valid_ref_accessors = []\n",
    "        for i in ref_suffixes:\n",
    "            i_prefix = '/'.join(i.split('/')[:-1])\n",
    "            \n",
    "            if str(data[i_prefix +'/DOI'])=='nan' or str(data[i_prefix +'/DOI'])=='No DOI':\n",
    "                try_delete(data,i)\n",
    "                try_delete(data,i_prefix)\n",
    "                try_delete(data,i_prefix+'/NOTES')\n",
    "            else:\n",
    "                data[i_prefix+'/DOI'] = data[i_prefix+'/DOI'].split(' ')[-1]\n",
    "                try_delete(data,i_prefix)\n",
    "                try_delete(data,i_prefix+'/NOTES')                \n",
    "                valid_ref_accessors.append(i_prefix)\n",
    "\n",
    "        refs = []\n",
    "        titles = []\n",
    "        for i in valid_ref_accessors:\n",
    "            doi = data[i+'/DOI']\n",
    "            title = get_title_from_doi(doi)\n",
    "            \n",
    "            refs.append(doi)\n",
    "            titles.append(title)\n",
    "            \n",
    "        refs = ';;'.join(refs)\n",
    "        titles = ';;'.join(titles)\n",
    "            \n",
    "        if not str(data[ctid]).startswith('CL:'):         \n",
    "            continue\n",
    "        \n",
    "        tissue_general = None\n",
    "        for i in tissue_suffixes[::-1]:\n",
    "            if data[i] in tissue:\n",
    "                tissue_general = data[i]\n",
    "                break\n",
    "\n",
    "        tissue_specific = None\n",
    "        for i in tissue_suffixes[::-1]:\n",
    "            if data[i] in tissue_original:\n",
    "                tissue_specific = data[i]\n",
    "                break\n",
    "\n",
    "        if tissue_general is None:\n",
    "            for i in tissue_suffixes:\n",
    "                if data[i].startswith(\"UBERON:\"):\n",
    "                    tissue_general=data[i]\n",
    "                    break\n",
    "                    \n",
    "        if tissue_specific is None:\n",
    "            for i in tissue_suffixes:\n",
    "                if data[i].startswith(\"UBERON:\"):\n",
    "                    tissue_specific=data[i]                    \n",
    "                    break\n",
    "                    \n",
    "        assert tissue_general is not None\n",
    "        assert tissue_specific is not None\n",
    "\n",
    "        for gene in genes:\n",
    "            label = str(data[gene_to_key[gene]+'/LABEL'])\n",
    "            if gene == label.upper() or label == 'nan':\n",
    "                label = get_gene_name(gene)\n",
    "\n",
    "\n",
    "            gene_dict = {\n",
    "                \"tissue_general\": tissue_general,\n",
    "                \"tissue_specific\": tissue_specific,\n",
    "                \"symbol\": gene,\n",
    "                \"name\": label,\n",
    "                \"publication\": refs,\n",
    "                \"publication_titles\": titles,\n",
    "                \"cell_type_ontology_term_id\": data[ctid]\n",
    "            }\n",
    "            hashed_dict = hash(json.dumps(gene_dict))\n",
    "            if hashed_dict not in seen:\n",
    "                parsed_table_entries.append(gene_dict)\n",
    "                seen.append(hashed_dict)\n",
    "                \n",
    "\n",
    "ts = list(set([i['tissue_general'] for i in parsed_table_entries]+[i['tissue_specific'] for i in parsed_table_entries]))\n",
    "\n",
    "tissues_by_id = {t: get_tissue_name(t) for t in ts}\n",
    "\n",
    "gene_infos = {}\n",
    "for entry in parsed_table_entries:\n",
    "    entry = entry.copy()\n",
    "    ct = entry['cell_type_ontology_term_id']\n",
    "    del entry['cell_type_ontology_term_id']\n",
    "    \n",
    "    a = gene_infos.get(ct,[])\n",
    "    entry['tissue_general'] = tissues_by_id.get(entry['tissue_general'],entry['tissue_general'])\n",
    "    entry['tissue_specific'] =tissues_by_id.get(entry['tissue_specific'],entry['tissue_specific'])\n",
    "    a.append(entry)\n",
    "    gene_infos[ct]=a\n",
    "\n",
    "\n",
    "json.dump(gene_infos,open('allCellTypeMarkerGenes.json','w'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
