# Contributing Data

CELLxGENE supports a rapidly growing single-cell data corpus because of generous contributions from researchers like you!

## Submission and Publication Process

1. Review the [Data Eligibility](#data-eligibility) criteria to ensure your data complies with these requirements
2. [Contact us] with a description of the data that you'd like to contribute to confirm that we will accept your data
3. Once confirmed, you send us files prepared according to the submission [Requirements](#dataset-requirements)
4. We upload to a private Collection where you can review
5. The submission can be revised, as needed
6. The data are made openly available when you are ready

## Dataset Requirements

### Data Eligibility

CELLxGENE supports most single-cell RNA-seq and ATAC-seq data, but a few types of data are **not** accepted at this time:

- drug screens
- cell lines
- species not on the [supported list](https://chanzuckerberg.github.io/single-cell-curation/latest-schema.html#organism_ontology_term_id)
- assays not on the [supported list](https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/census_accepted_assays.csv)
  - these additional assays are accepted:
    - expression data from paired (i.e. multi-modal) assays (e.g. 10x multiome, mCT-seq)
    - unpaired scATAC-seq gene activity matrices with fragments file

CELLxGENE continues to expand support for additional species and assays so please [contact us] if you are interested in submitting data not currently covered by the supported lists.

### Scale Constraints

CELLxGENE Discover sets the maximum per dataset file size for submissions to 50 GB. Additionally, datasets with more than 4.3 million cells can be submitted but will not visualized in CELLxGENE Explorer.

### Formatting Requirements

Include the following Collection metadata in your emails to describe your publication or study, all of which can be edited as titles, abstracts, etc.:

- **Collection information**:
  - Title
  - Description
  - Contact: a single name and email
  - Publication/preprint DOI: _optional_
  - URLs _optional_
    - any links to the corresponding raw sequence data, protocols, and other related data or resources
  - Consortia _optional_
    - one or more of those listed [here](https://github.com/chanzuckerberg/single-cell-data-portal/blob/main/backend/layers/common/validation.py#L12)

The full schema is documented [here](https://chanzuckerberg.github.io/single-cell-curation/latest-schema.html) but is summarized below. Each dataset needs the following information added to a single h5ad (AnnData 0.10) format file:

- **Dataset-level metadata in uns**:
  - **title**
  - **organism_ontology_term_id**
    - [NCBITaxon] See [the schema](https://chanzuckerberg.github.io/single-cell-curation/latest-schema.html#organism_ontology_term_id) for specific values
  - **batch_condition** _optional_
    - list of obs fields that define “batches” that a normalization or integration algorithm should be aware of
  - **default_embedding** _optional_
    - the obsm key associated with the embeddings you would like to be displayed in CELLxGENE Explorer by default
- **Data in .X and raw.X**:
  - raw counts are required
  - normalized counts are strongly recommended
  - raw counts should be in raw.X if normalized counts are in .X
  - if there is no normalized matrix, raw counts should be in .X
  - layer in .X is displayed in CELLxGENE Explorer, note that features in .X can be extensively filtered
- **Cell metadata in obs** (ontology term IDs MUST be the most specific term available from the specified ontology):
  - **donor_id**
    - free-text identifier that distinguishes the unique individual that data were derived from
    - `na` for cell line
  - **development_stage_ontology_term_id**
    - human [HsapDv]
    - mouse [MmusDv]
    - roundworm [WBls](https://www.ebi.ac.uk/ols4/ontologies/wbls/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FWBls_0000825?lang=en)
    - zebrafish [ZFS]
    - fruit fly FBdv descendent of [development stage](https://www.ebi.ac.uk/ols4/ontologies/fbdv/classes?obo_id=FBdv%3A00005259) or [age](https://www.ebi.ac.uk/ols4/ontologies/fbdv/classes?obo_id=FBdv%3A00007014)
    - other organsism [UBERON](https://www.ebi.ac.uk/ols4/ontologies/uberon/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FUBERON_0000105)
    - `unknown` if information unavailable
    - `na` for cell line
  - **sex_ontology_term_id**
    - `PATO:0000384` for male, `PATO:0000383` for female, `PATO:0001340` for hermaphrodite, or `unknown` if unavailable
    - `na` for cell line
  - **self_reported_ethnicity_ontology_term_id**
    - human [HANCESTRO]
    - multiple <code>&nbsp;||&nbsp;</code>-delimited terms may be used if more than one ethnicity is reported
    - `unknown` if information unavailable
    - other organisms `na`
    - `na` for cell line
  - **disease_ontology_term_id**
    - should describe any known disease thought to, or being tested to, have an impact on the measurement being taken, _not_ necessarily any known disease of the donor
    - [MONDO] or `PATO:0000461` for normal
    - multiple <code>&nbsp;||&nbsp;</code>-delimited terms may be used if appropriate
  - **tissue_type**
    - `tissue`, `organoid`, `primary cell culture` or `cell line`
  - **tissue_ontology_term_id**
    - should describe the sample used as input to the experiment, _not_ analysis-derived annotations
    - round worm [WBbt](https://www.ebi.ac.uk/ols4/ontologies/wbbt/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FWBbt_0005766) or [UBERON]
    - zebrafish [ZFA](https://www.ebi.ac.uk/ols4/ontologies/zfa/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FZFA_0100000?lang=en) or [UBERON]
    - fruit fly [FBbt](https://www.ebi.ac.uk/ols4/ontologies/fbbt) or [UBERON]
    - other organisms [UBERON]
    - cell line [CVCL](https://www.cellosaurus.org/)
    - primary cell culture [CL]
  - **cell_type_ontology_term_id**
    - should describe analysis-derived cell annotations
    - roundworm [WBbt](https://www.ebi.ac.uk/ols4/ontologies/wbbt/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FWBbt_0004017) or [CL]
    - zebrafish [ZFA](https://www.ebi.ac.uk/ols4/ontologies/zfa/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FZFA_0009000) or [CL]
    - fruit fly [FBbt](https://www.ebi.ac.uk/ols4/ontologies/fbbt/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FFBbt_00007002) or [CL]
    - other organisms [CL]
    - cell line may be all `na`
  - **assay_ontology_term_id**
    - [EFO]
  - **suspension_type**
    - `cell`, `nucleus`, or `na`
- **Embeddings in obsm**:
  - one or more two-dimensional embeddings, prefixed with 'X\_'
- **Features in var & raw.var (if present)**:
  - index is Ensembl gene ID
  - recommendation is that genes have not been filtered in order to maximize future data integration efforts
- **Additional standards for single-capture area Visium datasets** (largely aligns with [scanpy’s model](https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_visium.html), [this notebook](https://github.com/Lattice-Data/lattice-tools/blob/main/cellxgene_resources/curation_visium.ipynb) may be helpful to curate from Space Ranger outputs):
  - empty spots must be included
    - 4992 total observations for 6.5 mm capture areas
    - 14336 total observations for 11 mm capture areas
  - **obsm['spatial']**
  - **obs['array_row']**
  - **obs['array_col']**
  - **obs['in_tissue']**
  - **uns['spatial'][library_id]['images']['fullres']** _recommended_
    - fullres image that is _input_ to Space Ranger
  - **uns['spatial'][library_id]['images']['hires']**
    - hires image that is _output_ from Space Ranger
  - **uns['spatial'][library_id]['scalefactors']['spot_diameter_fullres']**
  - **uns['spatial'][library_id]['scalefactors']['tissue_hires_scalef']**
  - multiple-capture area Visium datasets are permitted if each capture area is also submitted individually
    - the additional Visium standards do not apply to these
- **Additional standards for single-puck Slide-seq datasets**:
  - **obsm['spatial']**
  - multiple-puck Slide-seq datasets are permitted if each puck is also submitted individually
    - obsm['spatial'] is not required for these
- **Additional ATAC-seq standards**
  - A fragments file is required for each unpaired scATAC-seq submission, _recommended_ for multi-modal submission
  - Follows the 5-column tabular format produced by [Cell Ranger](https://www.10xgenomics.com/support/software/cell-ranger-atac/latest/analysis/outputs/fragments-file)
  - Barcode values must match the values in the obs index of the corresponding AnnData object
  - [This notebook](https://github.com/Lattice-Data/lattice-tools/blob/main/cellxgene_resources/curation_fragments.ipynb) may be helpful to curate from Cell Ranger outputs

## Data Submission Policy

I give CZI permission to display, distribute, and create derivative works (e.g. visualizations) of this data for purposes of offering CELLxGENE Discover, and I have the authority to give this permission. It is my responsibility to ensure that this data is not identifiable. In particular, I commit that I will remove any [direct personal identifiers](https://docs.google.com/document/d/1sboOmbafvMh3VYjK1-3MAUt0I13UUJfkQseq8ANLPl8/edit) in the metadata portions of the data, and that CZI may further contact me if it believes more work is needed to de-identify it. If I choose to publish this data publicly on CELLxGENE Discover, I understand that (1) anyone will be able to access it subject to a CC-BY 4.0 license, meaning they can download, share, and use the data without restriction beyond providing attribution to the original data contributor(s) and (2) the Collection details (including Collection name, description, my name, and the contact information for the datasets in this Collection) will be made public on CELLxGENE Discover as well. I understand that I have the ability to delete the data that I have published from CELLxGENE Discover if I later choose to. This however will not undo any prior downloads or shares of such data.

[contact us]: mailto:cellxgene@chanzuckerberg.com
[CL]: https://www.ebi.ac.uk/ols/ontologies/cl
[EFO]: https://www.ebi.ac.uk/ols/ontologies/efo
[HANCESTRO]: https://www.ebi.ac.uk/ols/ontologies/hancestro
[HsapDv]: https://www.ebi.ac.uk/ols/ontologies/hsapdv
[MONDO]: https://www.ebi.ac.uk/ols/ontologies/mondo
[MmusDv]: https://www.ebi.ac.uk/ols/ontologies/mmusdv
[NCBITaxon]: https://www.ebi.ac.uk/ols4/ontologies/ncbitaxon
[UBERON]: https://www.ebi.ac.uk/ols/ontologies/uberon
[ZFS]: https://www.ebi.ac.uk/ols4/ontologies/zfs
