# Roadmap

## Roadmap Vision

We are building cellxgene, a data publishing and exploration platform that maximizes the reusability of published data by standardizing it and making it findable and accessible. Cellxgene will provide cell-centric data access APIs for computational reuse and interactive graphical interfaces so that biologists and physicians can ask questions directly of the data.

## Roadmap Themes

Our technology roadmap aims to:

1. Make it easier to find data and evaluate if existing data can be used to answer novel research questions
2. Make it easier for biologists, physicians, and computational biologists to use data to answer questions

## Features

We are planning to develop the following features on the Data Portal application. The Status column has one of three statuses:

- Development: Our engineering team is currently working hard to deliver this feature.
- Discovery: We're in the process of defining requirements and prototyping this feature.
- Not Started: Self explanatory! We haven't started Discovery yet.

<table>
  <thead>
    <tr>
      <th style="text-align:left">
        <p>Expected</p>
        <p>Release</p>
      </th>
      <th style="text-align:left">Feature</th>
      <th style="text-align:left">Description</th>
      <th style="text-align:center">Supports Theme</th>
      <th style="text-align:left">Status</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align:left">2021</td>
      <td style="text-align:left">
        <a href="https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/211">
          Find datasets
        </a>
      </td>
      <td style="text-align:left">
        Users can filter by metadata to find relevant datasets
      </td>
      <td style="text-align:center">1</td>
      <td style="text-align:left">Development</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">
        <a href="https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/104">
          Gene sets
        </a>
      </td>
      <td style="text-align:left">
        Users can publish gene sets containing marker genes, pathways, or
        mechanisms along with datasets. Scientists coming to cellxgene can
        download and explore data with published gene sets
      </td>
      <td style="text-align:center">1</td>
      <td style="text-align:left">Discovery</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">
        Differential expression scale and speed improvements
      </td>
      <td style="text-align:left">
        Differential expression runs on million cell scRNA-seq datasets in tens
        of seconds
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Discovery</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">Where&apos;s my gene?</td>
      <td style="text-align:left">
        Users can view the expression of gene(s) across cell types in our public
        data corpus spanning all available human tissues
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Discovery</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">
        <a href="https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/57">
          Protein &amp; splicing data support
        </a>
      </td>
      <td style="text-align:left">
        Users can publish and explore datasets that contain multiple data
        modalities (i.e. Protein &amp; mRNA).
      </td>
      <td style="text-align:center">1, 2</td>
      <td style="text-align:left">Discovery</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">Download selected cells from the corpus</td>
      <td style="text-align:left">
        Users can select cells that match metadata in our schema and download a
        concatenated set of cells from our UI or in their local compute
        environment using our API
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022</td>
      <td style="text-align:left">
        Integrate selections of cells and explore the results
      </td>
      <td style="text-align:left">
        Users can select cells to integrate and then explore them using
        cellxgene&apos;s visual tools
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022+</td>
      <td style="text-align:left">Annotate and persist integrated datasets</td>
      <td style="text-align:left">
        Once a user integrates multiple datasets, they can create cell labels
        and gene sets. The integrated dataset, cell labels, and gene sets will
        be downloadable, and will persist the next time they log in
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022+</td>
      <td style="text-align:left">Integrate my dataset with public datasets</td>
      <td style="text-align:left">
        Users can (privately) upload their dataset, select [cell types or
        datasets] to integrate with from the data corpus, predict labels and
        annotations for their data, and explore or download the result
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022+</td>
      <td style="text-align:left">Gene set enrichment analysis</td>
      <td style="text-align:left">
        Users can submit a list of genes and receive gene sets that are likely
        matches. Users can view the gene sets that are most differentially
        expressed between two populations of cells
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022+</td>
      <td style="text-align:left">Where&apos;s my pathway?</td>
      <td style="text-align:left">
        Users can view expression of predefined pathways (expressed as a gene
        set mean) for cell types across the data corpus
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
    <tr>
      <td style="text-align:left">2022+</td>
      <td style="text-align:left">Gene expression based indexing</td>
      <td style="text-align:left">
        When selecting cells to integrate or download, users can refine their
        selection with gene-expression based queries in addition to using
        metadata values. For example, users could select only T-cells that
        express CD4 and FOXP3
      </td>
      <td style="text-align:center">2</td>
      <td style="text-align:left">Not Started</td>
    </tr>
  </tbody>
</table>

## Please Note

- Our roadmap is subject to change based on community feedback and updates to our strategy.
- We update our roadmap every six months, this represents our best assessment of our future plans.
- We are fairly confident in delivering features in the short term. Longer term features we are less certain about what features we will deliver but are included to illustrate our vision for cellxgene.

## Feedback

We welcome your feedback! Please see the [Contact Us](contact-us.md) section for how to get in touch.
