# Find Marker Genes

Genes that are highly enriched in a particular group of cells can be used as "markers" for that population. Marker genes are commonly used to identify cell types in single-cell RNA-seq data. In Gene Expression, we can use the Find Marker Genes tool to identify genes that are specific to a particular cell type relative to other cell types in its tissue.


## Tutorial

First, let's select the lung tissue in Gene Expression. There is an info button next to each cell type label. Clicking the button opens a sidebar with the top 25 marker genes for that cell type:
<Image src={"/doc-site/4-2_plasma-cell.png"} height={250} />

The marker genes are ranked by their Marker Score (see *Algorithm* for more details). In the sidebar, we can click the `Add to Dot Plot` button to add these genes to the heatmap. 
<Image src={"/doc-site/4-2_add-dotplot.png"} height={250} />

Clicking the `Copy` button copies a comma-separated list of the marker genes to the clipboard. 
<Image src={"/doc-site/4-2_copy-button.png"} height={250} />

To minimize the sidebar, click the chevron in the upper-right corner of the sidebar.
<Image src={"/doc-site/4-2_close-button.png"} height={250} />

Clicking on the info button next to another cell type will replace the marker gene sidebar with the marker genes for the new cell type. 

## Algorithm
For a selected cell type in a tissue,
 - Use Welch's t-test to compare the gene expressions in the selected cell type to each other cell type in the tissue.
    - Record the effect size for each gene in each comparison.
 - For each gene, calculate the 5th percentile of the effect sizes across all comparisons. This is the marker score.
    - Let $T_t$ be the set of cell types $\{c_1,\ldots,c_n\}$ in tissue $t$. The marker score ($S_{c_ig_j}$) for gene $g_j$ in cell type $c_i$ is $S_{c_ig_j}=p_5(\{e_{c_kg_j} : c_k \in T_t, c_k \neq c_i\})$ where $p_5$ is the 5th percentile operator and $e_{c_kg_j}$ is the effect size of the t-test comparing $g_j$ in $c_i$ to $c_k$.
 - Sort the genes by their marker score in descending order and return the top 25.

## Welch's t-test
Welch's t-test is a non-parametric test for comparing the means of two independent samples with unequal sample sizes and variances. For a particular gene,
 - Let two groups of cells be $c_1$ and $c_2$. Calculate the following values:
    - $m_1$ and $m_2$ are the average expression of the gene in $c_1$ and $c_2$ respectively.
    - $s_1$ and $s_2$ are the standard deviations of the gene in $c_1$ and $c_2$ respectively.
    - $n_1$ and $n_2$ are the number of cells in $c_1$ and $c_2$ respectively _excluding cells from datasets that did not measure the gene_.
 - Calculate the effect size (Cohen's d):
    - $e = \frac{m_1 - m_2}{\sqrt{\frac{(n_1 - 1)s_1^2 + (n_2 - 1)s_2^2}{n_1 + n_2 - 1}}}$.    
 - Calculate the t-statistic $t$:
    - $t = \frac{m_1 - m_2}{\sqrt{\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}}}$.
 - Approximate degrees of freedom using Satterthwaite formula:
    - $df = \frac{\left(\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}\right)^2}{\frac{\left(\frac{s_1^2}{n_1}\right)^2}{n_1 - 1} + \frac{\left(\frac{s_2^2}{n_2}\right)^2}{n_2 - 1}}$.
 - Calculate the p-value using the t-distribution ($P(T >t)$).

## Caveats
 - It is important to note that some methodological decisions were made to balance accuracy with efficiency and scalability. For example, we use a t-test to perform differential expression, which is a simple and fast test. However, it may not be as accurate as more sophisticated (and computationally intensive) tests, such as the Wilcoxon rank-sum test. 
 - While differential expression is typically performed on raw counts for single-cell RNA sequencing data, we opted to use the <NextLink href="04__Analyze%20Public%20Data/4_2__Gene%20Expression%20Documentation/4_2_3__Gene%20Expression%20Data%20Processing">rankit-normalized</NextLink> values in order to identify marker genes that corroborate the gene expressions shown in the dot plot.
 - For the beta, we opted to not report the p-values for each gene. Though we can calculate approximate p-values using the t-statistic and degrees of freedom calculated in *Welch's t-test*, it is difficult to accurately aggregate the p-values for each gene across all comparisons. The most reliable method involves repeatedly permuting the data to generate a null distribution, but this is computationally intensive.
 - Currently, marker genes are only calculated for cell types in tissues using only _healthy_ cells. Applying secondary filters to the data (like disease, ethnicity, etc.) does not affect the results. Enabling dynamic calculation of marker genes for arbitrary populations of cells in arbitrary subsets of the data may be a direction for future development.
 - Ideal marker genes for a particular cell type are expressed in all of its cells and not expressed anywhere else: they have binary expression patterns. In reality, almost no genes have truly binary expression patterns. Instead, they have genes that are statistically enriched in their cells relative to all other cell types. Additionally, genes may be good markers for a cell type in one context (e.g. a tissue) but not another (e.g. the entire human body). 
 - Finally, it may be difficult to identify any good markers for some cell types, especially if they have a small number of cells or are compared to many similar cell types. To account for the former scenario, marker genes are not displayed for cell types with fewer than 25 cells. The latter scenario is particularly relevant in the blood as it contains many closely-related cell types. As a result, we have temporarily disabled the find marker genes feature for cell types in blood.