# CZ CELLxGENE Explorer Tutorials

The CELLxGENE Explorer is a visual tool linked to each dataset hosted by CELLxGENE Discover. Explorer enables analysis and interrogation of individual single cell data sets. This section describes Explorer's capabilities.

Just getting started with the tool? We recommend going through these tutorials in order to learn how to use the tool.

Returning for a quick refresh? We’ve identified the key concepts below the tutorial so you can skim through and find what you’re looking for.

You can follow along with any of these tutorials by launching an instance of the CELLxGENE Explorer with the Tabula Sapiens using [this CELLxGENE Explorer link](https://cellxgene.cziscience.com/e/53d208b0-2cfd-4366-9866-c3c6114081bc.cxg/). For more information on Tabula Sapiens, you can refer to[ the preprint on biorxiv](https://www.biorxiv.org/content/10.1101/2021.07.19.452956v1).

## CZ CELLxGENE Explorer Interface

CELLxGENE Explorer's user interface organizes single cell data similarly to how it is organized in single cell data formats. The left hand side displays categorical and numerical sample metadata. The right hand side is a space for displaying features such as genes and gene sets. The center displays the embedding, where each cell is a point. UMAP and tSNE are common embeddings, which place cells based on their local distances in gene expression space. Cells from spatial data can also be displayed using each cell's (x, y) coordinates.

**Key Concepts**: User Interface Explanation

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1W4YimCRcKgjkLUBWd4bqcy-wDxZQXOmDCvDThugFE54/embed?start=false&loop=false&delayms=3000" />

## Examining Categorical Metadata

Categorical metadata (such as tissue of origin or cell type) can be used in a number of ways within Explorer including coloring embedding plots (i.e. color UMAP by cell type), looking at cell counts, within a categorical metadata field, making selections of cells or viewing the interaction between different categorical metadata fields.

**Key Concepts**: Categorical Metadata, Selecting Cells by Category (i.e. cell type), Interaction Between Categorical Metadata Fields

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/13B1ThLYyvdJzTzrw_MHSu1FgxwLxE0JDmMKNQNYt0TI/embed?start=false&loop=false&delayms=3000" />

## Find Cells Where a Gene is Expressed

Numerical metadata (such gene expression features or QC metrics such as number of genes) can be examined on the embedding plot and be used to filter and select cells. Additionally tools like the clip tool can give us control over how these attributes are displayed on the embedding plot.

**Key Concepts**: Numerical Metadata, Cell Filtering and Selection, Interaction Between Numerical Metadata, Categorical Metadata Fields

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1Z9mD5SK-6nI8pogQEclI9iYDI4vI_zoigblAy0d3tvM/embed?start=false&loop=false&delayms=3000" />

## Selecting and Subsetting Cells

Explorer allows for the complex selection of cells via selection directly on the embedding, gene expression cutoffs, and based on categorical metadata attributes.

**Key Concepts**: Categorical Metadata Selection, Numerical Metadata Selection, Complex Selection (combining selection methods)

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1UvnH7yVB66-xUWIcWsPJ_U0cB17CjhRbWtQHnP59WiA/embed?start=false&loop=false&delayms=3000" />

## Compare Expression of Multiple Genes

Explorer allows you to compare the expression of multiple genes via bivariate plots.

**Key Concepts**: Gene Expression, Co-expression, Cell Selection, Subsetting

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1v5J4ggWK7mI4DyEibqbBSLGt2CIFDNjrWuKSxmz-MCQ/embed?start=false&loop=false&delayms=3000" />

## Using Gene Sets to Learn About Cell Population Functional Characteristics

Explorer allows you to examine groups of genes via the gene sets feature.

**Key Concepts**: Gene Expression, Co-expression, Cell Selection, Subsetting

_Listed below is a comma separated gene set list for use with this tutorial._

ACAA1, ACAA2, ACADL, ACADM, ACADS, ACADSB, ACADVL, ACAT1, ACAT2, ACOX1, ACOX3, ACSL1, ACSL3, ACSL4, ACSL5, ACSL6, ADH1A, ADH1B, ADH1C, ADH4, ADH5, ADH6, ADH7, ALDH1B1, ALDH2, ALDH3A2, ALDH7A1, ALDH9A1, CPT1A, CPT1B, CPT1C, CPT2, CYP4A11, CYP4A22, ECHS1, ECI1, ECI2, EHHADH, GCDH, HADH, HADHA, HADHB

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1pv5In5uD_v1Sy6vSKHPvM7e7ViCvftOALelX3wbKAp0/embed?start=false&loop=false&delayms=3000" />

## Find Marker Genes

Explorer allows you to find marker genes between selected cell populations.

**Key Concepts**: Gene Expression, Differential Expression, Cell Selection, Subsetting

<EmbeddedGoogleSlides src="https://docs.google.com/presentation/d/1sNDshrmj7uBXEYtDNV0kKYAa3_Y-cWuiAus64kIRlxM/embed?start=false&loop=false&delayms=3000" />

Note: You can find more information <NextLink href="/docs/04__Analyze%20Public%20Data/4_3__Algorithms%20and%20Ontologies#differential-expression">here</NextLink> about how our differential expression is calculated. In brief, we use a Welch's t-test. While we are aware that single cell data does not always meet the assumptions imposed by this test, we utilize it because it performs well at identifying the _most_ differentially expressed genes, and this is what our feature returns.
