# Automated Annotation of Single Cell Data (EXPERIMENTAL)

CZ CELLxGENE Annotate enables label transfer and reference mapping from expertly curated reference atlases to your newly generated and QC'd single cell datasets. Your data can be annotated using CELLxGENE Desktop along with any of our available label transfer and reference mapping models. CELLxGENE annotation augments your dataset with the following new information:
- transferred cell type labels (a.k.a. the most probable cell type)
- confidence scores (certainty measure of the prediction of cell type identity - takes on a value between 0 and 1)
- a projection of the query dataset into the reference embedding, and a newly computed UMAP based on that projection

After you successfully complete the automatic annotation workflow, you can explore and refine your predictions using [standard features in CELLxGENE](https://cellxgene.cziscience.com/docs/04__Analyze%20Public%20Data/4_1__Hosted%20Tutorials) such as:
- visualizing clusters by predicted cell type
- plotting marker genes
- performing differential expression
- refining cell type labels by creating custom annotations

**Disclaimer:** The Annotation feature in CELLxGENE is new and experimental. We intend for the results generated by the workflow to be a starting point for annotating your single cell dataset. If you run into issues completing this workflow, we would love to hear about your experience. Feel free to reach out to [cellxgene@chanzuckerberg.com](cellxgene@chanzuckerberg.com) to get help running the workflow or to forward constructive feedback. In the meantime, in order to obtain the best results, we make the following recommendations:
- use this predictive annotation workflow with data generated by 10X platforms
- use the tissue model which is most closely related to the tissue profiled in your single cell experiment

## Installation of CELLxGENE[annotate]


You can install the latest experimental release of CELLxGENE to gain access to the automatic annotation subcommand. It is highly recommended to perform installation in a clean virtual environment (such as conda or venv) with python version 3.8+. Find examples of virtual environment setup using venv and conda below

### venv setup

For an overview of `venv`, refer [here](https://docs.python.org/3/tutorial/venv.html)

```
brew install pyenv  # used by MLflow

virtualenv --python python3.8 /path/to/venv

. /path/to/venv/bin/activate

```

### conda setup

To install `conda`, refer [here](https://www.anaconda.com/products/distribution)

```
conda create -n cellxgene 'python=3.9'

conda activate cellxgene
```

### pip install
Once your virtual environment is installed and activated, install the `annotate` feature using pip using the following command:

`pip install 'cellxgene[annotate]'`

You can run the following to test the success of your installation:

`cellxgene annotate -h` or `cellxgene annotate --help`

## CELLxGENE Annotation Workflow

See our quickstart guide to  quickly see if you can run through the pipeline or check out the more detailed sections below for more detailed descriptions of the workflow. Both assume that you have properly set up a clean virtual environment to work from.

### Quickstart

Depending on your environment manager, execute one of the following quickstart commands:

#### venv
```
mkdir annotations #create a directory to download data, run annotation pipeline and store annotated object

cd annotations #navigate to newly created directory

wget https://github.com/chanzuckerberg/cellxgene/raw/main/example-dataset/pbmc3k.h5ad # retrieve sample dataset

cellxgene annotate -i minilung.h5ad  -m https://cellxgene-annotation-public.s3.us-west-2.amazonaws.com/cell_type/models/hlca_ref_20220728145305_conda.zip -o miniLungAnnotated.h5ad --gene-column-name feature_name --mlflow-env-manager conda --no-use-gpu # generate annotations

cellxgene launch miniLungAnnotated.h5ad #launch cellxgene explorer
```

#### conda

```
mkdir annotations #create a directory to download data, run annotation pipeline and store annotated object

cd annotations #navigate to newly created directory

wget https://github.com/chanzuckerberg/cellxgene/raw/main/example-dataset/pbmc3k.h5ad ### TODO - wget command for s3 hosted minilung dataset

cellxgene annotate -i minilung.h5ad  -m https://cellxgene-annotation-public.s3.us-west-2.amazonaws.com/cell_type/models/hlca_ref_20220728145305_conda.zip -o miniLungAnnotated.h5ad --gene-column-name feature_name --mlflow-env-manager conda --no-use-gpu # generate annotations

cellxgene launch miniLungAnnotated.h5ad #launch cellxgene explorer
```

### Step 0: Annotation workflow inputs

CELLxGENE's automated annotation has two essential inputs:
- a query dataset to be annotated (structured as an `anndata` object (read more about the `anndata` object [here](https://anndata.readthedocs.io/en/latest/)))
- a model that will be used to generate the annotations 

#### Query data
You should create an `anndata` object (written to an `h5ad` file) that is compatible with the `annotate` subcommand:
- contains raw counts in the `adata.X`, `adata.raw.X`, or in a specified layer stored in `adata.layers`
- meets the additional [requirements](https://cellxgene.cziscience.com/docs/05__Annotate%20and%20Analyze%20Your%20Data/5_3__Preparing%20Data) for launching and visualizing a dataset with CELLxGENE desktop
  - note that the input object does not require a visualization embedding and that one will be generated as a part of running the annotation pipeline
  - the embedding that results from running the annotation pipeline will be stored as `adata.obsm['cxg_cell_type_umap']`, however you can use the command line options `--annotation-prefix`, `--annotation-type`, and `--run-name` to manipulate the name of the resulting embedding 

**Generating annotations from 10X cellranger outputs**

[10X cellranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov) is one of the standard pipelines for quantifying gene expression values (generating the count matrix). While the `annotate` subcommand does not currently offer the ability to read cellranger outputs directly, it is relatively easy to create a compatible `anndata` object using `scanpy`. Here are some resources to get you started:

- [ScanPy PBMC 3K clustering tutorial](https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html)
- [scanpy.read_10x_mtx()](https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_10x_mtx.html)
- [scanpy.read_10x_h5()](https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_10x_h5.html)

It is important that once you have constructed your object, that you perform initial QC filtering of low quality cells from the dataset. See this [link](https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html) for more details (specifically, the section entitled 'Basic Filtering').


#### Prediction model
You can download models to perform prediction for the following tissues:
- [TODO: lung](www.google.com)
- [TODO: immune](www.google.com)



### Step 1: Running `cellxgene annotate`

After installing the latest version of CELLxGENE and the annotate subcommand, you can run a standard call to the command like so (see `cellxgene annotate --help` for a full list of available options):

`cellxgene annotate  --input-h5ad-file ./pathTo/data.h5ad --model-url modelURL --output-h5ad-file ./pathTo/annotatedData.h5ad`

In the call of the `annotate` subcommand above, we specify three arguments
- the local path to your query dataset (in `anndata` format)
- the url of the tissue model which you like to use to perform reference mapping (for a list of available reference models, refer to the [section above](#prediction-model))
- the local path where `cellxgene annotate` will write a new `anndata` object that includes the predicted annotations

**Note:** the first time use of a given model will:
- trigger a one-time download of the model, stored locally for repeated use (in .model_cache by default)
- take some extra time to setup an (internal) python env in which to run the model

Once your call to `cellxgene annotate` completes, you will have the following new information in your `anndata` object:

- transferred annotations/labels [`categorical`] (i.e. cell type) - these will be stored under `adata.obs['cxg_cell_type_predicted']` by default
- confidence scores [`numeric`] - stored under `adata.obs['cxg_cell_type_predicted_uncertainty']` by default
- a mapping into the reference embedding space - stored under `adata.obsm['cxg_cell_type'], by default`
- a UMAP layout for the query data in this reference embedding space - stored under `adata.obsm['cxg_cell_type_umap']` by default
- metadata about the model used - stored under `adata.uns['cxg_cell_type_predictions']['model_url']` by default

The `cellxgene annotate` subcommand offers a number of options to customize how the workflow is run and to adapt to different scenarios. You can display the full list of available options like so:

`cellxgene annotate -h`


### Step 2: Explore and Refine

After you have successfully completed the annotation workflow, you are now ready to explore your results in CELLxGENE Desktop!

You can launch your annotated `anndata` object in the CELLxGENE explorer like so:

`cellxgene launch ./pathTo/annotatedData.h5ad`

You will now have access to the standard set of CELLxGENE features to explore your single cell dataset. Here are some initial steps that you might take to start exploring your annotated dataset:

- color by predicted cell type
- color by prediction confidence score
- calculate differential expression between two annotated populations to see which genes distinguish them from each other
- create custom annotations to further refine or correct the predicted annotations

For a reference on how one might refine predicted annotations, you can refer to this [Nature Tutorial](https://www.nature.com/articles/s41596-021-00534-0).

### Useful Links
- [10X cellranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov)
- [Seurat](https://satijalab.org/seurat/)
- [ScanPy](https://scanpy.readthedocs.io/en/stable/)
- [annotation](https://cellxgene.cziscience.com/docs/05__Annotate%20and%20Analyze%20Your%20Data/5_4__Annotating%20Data)
- [Tutorial: guidelines for annotating single-cell transcriptomic maps using automated and manual methods](https://www.nature.com/articles/s41596-021-00534-0)
