# Quick Start

### Overview

Cellxgene Desktop Explorer is a tool in the cellxgene suite that enables collaborative exploratory analysis of single cell data and private data sharing.

This page describes how to quickly get started with exploratory analysis on your local computer. Those interested in hosting cellxgene to enable your collaborators to explore data from their browser without needing to download the data file or install cellxgene should see [Self-Host Cellxgene](self-hosting/).

### Quick Start

[To install cellxgene](install.md) you need Python 3.6+. We recommend installing cellxgene into a [conda](install.md#using-a-conda-environment) or [virtual environment](install.md#using-a-virtual-environment).

Install the package, then launch cellxgene with an example [AnnData](https://anndata.readthedocs.io) file:

```text
pip3 install cellxgene
cellxgene launch https://cellxgene-example-data.czi.technology/pbmc3k.h5ad
```

## Example datasets

The following datasets are available on the web and can be opened directly using `cellxgene launch`

### Peripheral blood mononuclear cells

| Source                                                                                            | Cells | File Size | Raw Data | Processing                                                                                                                     |
| :------------------------------------------------------------------------------------------------ | :---- | :-------- | :------- | :----------------------------------------------------------------------------------------------------------------------------- |
| [10X genomics](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k) | 2638  | 19MB      | Link     | [Notebook](https://github.com/chanzuckerberg/cellxgene-compbio-methods/blob/master/dataset-processing/pbmc3k-processing.ipynb) |

```text
cellxgene launch https://cellxgene-example-data.czi.technology/pbmc3k.h5ad
```

### Tabula muris

| Source                                                                | Cells | File Size | Raw Data                                                                                                                                                      | Processing                                                                                                                           |
| :-------------------------------------------------------------------- | :---- | :-------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------ | :----------------------------------------------------------------------------------------------------------------------------------- |
| [bioRxiv, CZBiohub](https://www.biorxiv.org/content/10.1101/237446v2) | 45423 | 174MB     | [Link](https://figshare.com/projects/Tabula_Muris_Transcriptomic_characterization_of_20_organs_and_tissues_from_Mus_musculus_at_single_cell_resolution/27733) | [Notebook](https://github.com/chanzuckerberg/cellxgene-compbio-methods/blob/master/dataset-processing/tabula-muris-processing.ipynb) |

```text
cellxgene launch https://cellxgene-example-data.czi.technology/tabula-muris.h5ad
```

### Tabula muris senis

| Source                                                                | Cells | File Size | Raw Data                                                             | Processing                                                   |
| :-------------------------------------------------------------------- | :---- | :-------- | :------------------------------------------------------------------- | :----------------------------------------------------------- |
| [bioRxiv, CZBiohub](https://www.biorxiv.org/content/10.1101/661728v1) | 81478 | 3.9GB     | [Link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132040) | [Notebook](https://www.biorxiv.org/content/10.1101/661728v1) |

```text
cellxgene launch https://cellxgene-example-data.czi.technology/tabula-muris-senis.h5ad
```
