# Features

[Cellxgene Desktop](../desktop/quick-start.md) and [Data Portal](../portal/data-portal.md) have common UI that enables exploration and interrogation of single cell data. This section describes cellxgene's data exploration features.

> While most features are shared, [Annotation](../desktop/annotations.md) is only available on Cellxgene Desktop.

## Cross-filter cells and color by metadata

Categorical metadata can be used not only to color cells in the embedding scatterplot, but also to examine the representation of a particular metadata across other metadata fields. In the example below, check out how we can:

1. Select all the cells that have the value "marrow" in the tissue category
2. Subset these cells so that only the selection is displayed
3. Expand another metadata category \("subtissue"\) to reveal how cells from the marrow are distributed across different sub-tissues

![](../.gitbook/assets/crossfilter.gif)

## ![](../.gitbook/assets/image%20%282%29.png) Compare groups of cells with differential expression

Differential expression can be run between selections of two different groups of cells. In the media below, these steps are to:

1. Select cells for the first group \("B Cells"\)
2. Select cells for the second group \("granulocytes"\)
3. Click the differential expression icon
4. The top 10 differentially expressed features will appear on the right hand side-bar

![](../.gitbook/assets/diffexp.gif)

> **Note**: In the hosted explorer present on the cellxgene data portal, the differential expression feature has a limit of 50,000 cells \(i.e. the sum of the number of cells in group 1 and group 2 cannot exceed 50,000\). You can workaround this by only selecting subsamples of large clusters. We are currently working on ways to scale this calculation in the hosted setting.

> **Note**: You can find more information [here](algorithms.md#differential-expression) about how our differential expression is calculated. In brief, we use a Welch's t-test. While we are aware that single cell data does not always meet the assumptions imposed by this test, we utilize it because it performs well at identifying the _most_ differentially expressed genes. This enables rapid exploration without sacrificing accuracy.

## See how metadata and gene expression break down across different categories

Cellxgene can give you information on how different categorical and continuous metadata are distributed within any metadata field on the left-hand sidebar. To accomplish this, one simply needs to:

1. Color the cells by a desired categorical or continuous metadata field
2. Expand another categorical metadata field to examine either the cross-occupancy \(categorical metadata\) or distribution \(continuous metadata\) within that newly expanded field

![](../.gitbook/assets/category-breakdown.gif)

One could imagine using this feature in a scenario where it is expected that certain cell types are only present in a particular disease state.

## Find cells where a gene is expressed

Cellxgene provides functionality to select cells based on gene expression \(i.e. setting a numeric expression threshold or range\). To accomplish this, one can:

1. Search for a gene of interest using the search dialog in the upper right hand corner
2. Color by the gene which you would like to gate by
3. Brush over the histogram of gene expression in the right hand side bar to set the minimum and maximum thresholds for gene expression

![](../.gitbook/assets/gene-expression.gif)

## Compare expression of multiple genes

Cellxgene, allows for the display of bi-variate scatter plots which results in FACS like plots that can be used to examine expression of custom combinations of genes.To examine and compare the expression of multiple genes in cellxgene, one can:

1. Select a gene to plot on the X axis
2. Select a gene to plot on the Y axis

![](../.gitbook/assets/compare-genes.gif)

You can also also color by one of these genes to view expression in either the embedding plot of the bi-variate scatter plot. Additionally, you can still make manual selections on the embedding plot, while the bi-variate plot is active \(in the example above, we select a particular cluster where we want to view the expression relationship between the first gene and the second gene\)
